src/pipeline/cli_pipeline.c

```diff
@@ -0,0 +1,703 @@
+		rc = daos_sgls_copy_data_out(api_args->sgl_keys, 1, &pro->pro_sgl_keys, 1);
+		if (rc != 0)
+			D_GOTO(out, rc);
+	}
+	if (api_args->recx_size != NULL && pro->pro_recx_size.ca_count > 0) {
+		/** copying records' size */
+		memcpy((void *)api_args->recx_size, (void *)pro->pro_recx_size.ca_arrays,
+		       sizeof(*api_args->recx_size) * pro->pro_recx_size.ca_count);
+	}
+	if (pro->pro_sgl_recx.sg_nr_out > 0) {
+		/** copying record data (akeys' values) */
+		rc = daos_sgls_copy_data_out(api_args->sgl_recx, 1, &pro->pro_sgl_recx, 1);
+		if (rc != 0)
+			D_GOTO(out, rc);
+	}
+	for (i = 0; i < nr_agg; i++) {
+		/** copying aggregation buffers */
+		double             *src, *dst;
+		daos_filter_part_t *part;
+		char               *part_type;
+		size_t              length_part_type;
+
+		dst = (double *)api_args->sgl_agg->sg_iovs[i].iov_buf;
+		src = (double *)pro->pro_sgl_agg.sg_iovs[i].iov_buf;
+
+		if (first_ever_cb(api_args->anchor, cb_args->shard)) {
+			/**
+			 * This is the first time ever that this callback is executed for
+			 * this particular pipeline run.
+			 */
+			*dst = *src;
+			continue;
+		}
+
+		part             = api_args->pipeline->aggr_filters[i]->parts[0];
+		part_type        = (char *)part->part_type.iov_buf;
+
+		length_part_type = 20; /**
+					* XXX: we can do this because all function names
+					* are the same length. Change this in the future
+					* if needed.
+					*/
+
+		if (!strncmp(part_type, "DAOS_FILTER_FUNC_SUM", length_part_type) ||
+		    !strncmp(part_type, "DAOS_FILTER_FUNC_AVG", length_part_type)) {
+			*dst += *src;
+		} else if (!strncmp(part_type, "DAOS_FILTER_FUNC_MIN", length_part_type)) {
+			if (*src < *dst)
+				*dst = *src;
+		} else if (!strncmp(part_type, "DAOS_FILTER_FUNC_MAX", length_part_type)) {
+			if (*src > *dst)
+				*dst = *src;
+		}
+		api_args->sgl_agg->sg_iovs[i].iov_len = sizeof(double);
+	}
+	if (nr_agg > 0)
+		api_args->sgl_agg->sg_nr_out = nr_agg;
+
+	*api_args->nr_kds  = pro->pro_nr_kds;
+	*api_args->nr_iods = pro->pro_nr_iods;
+
+	if (api_args->stats != NULL) {
+		/** user wants stats */
+		if (first_ever_cb(api_args->anchor, cb_args->shard)) {
+			*api_args->stats = pro->stats;
+		} else {
+			api_args->stats->nr_objs += pro->stats.nr_objs;
+			api_args->stats->nr_dkeys += pro->stats.nr_dkeys;
+			api_args->stats->nr_akeys += pro->stats.nr_akeys;
+		}
+	}
+
+	/** anchor should always be updated at the end */
+	*api_args->anchor = pro->pro_anchor;
+
+out:
+	crt_req_decref(rpc);
+	tse_task_list_del(task);
+	tse_task_decref(task);
+
+	if (ret == 0) /** XXX: see obj_retry_error(int err) when retry I/O is implemented */
+		ret = rc;
+	return ret;
+}
+
+static void
+shard_pipeline_set_buffers_to_zero(d_sg_list_t *sgl)
+{
+	uint32_t i;
+
+	if (sgl != NULL && sgl->sg_iovs != NULL) {
+		for (i = 0; i < sgl->sg_nr; i++)
+			sgl->sg_iovs[i].iov_len = 0;
+		sgl->sg_nr_out = 0;
+	}
+}
+
+#define KDS_BULK_LIMIT	128
+
+static int
+shard_pipeline_run_task(tse_task_t *task)
```

Comment:
@wangdi1 Well...... Object API is not fully factorized, and a lot of common functionality between object and pipeline has to be replicated because object API functions (1) do more things that what I needed, and (2) uses internal object data structures and variables that can't be accessed outside the object directory.

When we started development, there was some discussions about where to put the pipeline code. One option was to put all the functionality within the Object API (make pipeline acceleration a feature of the Object API, instead of a new API). I think we decided against that, but I don't remember why. Maybe @mchaarawi can shed more light about this. In any case, I basically went over the object api to see how the code should be done and more or less replicate functionality but with the peculiarities of the pipeline thing.